File: JAVA/Largest Rectangle in Histogram.java
public class Solution {
    public int largestRectangleArea(int[] height) {
        int[] h = Arrays.copyOf(height, height.length+1);
        Stack<Integer> stack = new Stack<Integer>();
        int max = 0;
        int i = 0;
        while (i < h.length)
        {
            if (stack.isEmpty() || h[i] >= h[stack.peek()])
            {
                stack.push(i);
                i++;
            }
            else
            {
                int t = stack.pop();
                int tmp = h[t] * (stack.isEmpty() ? i : (i - stack.peek() - 1));
                max = Math.max(tmp, max);
            }
        }
        return max;       
    }
}

File: Largest Rectangle in Histogram.cpp
class Solution {
public:
    int largestRectangleArea(vector<int> &height) {
        if (height.size() == 0) return 0;
        height.push_back(-1);
        stack<int> stk;
        int result = 0;
        int i = 0;
        while(i < height.size()) {
            if (stk.empty() || height[stk.top()] < height[i]) {
                stk.push(i);
                i++;
            } else {
                int j = stk.top();
                stk.pop();
                int area = height[j] * (i - 1 - (stk.empty() ? -1 : stk.top()));
                result = max(result, area);
            }
        }
        height.pop_back();
        return result;
    }
};
